Source/Glide64/3dmath.cpp

```diff
@@ -151,48 +151,119 @@ void calc_sphere (VERTEX *v)
 float DotProductC(register float *v1, register float *v2)
 {
     register float result;
-    result = v1[0]*v2[0] + v1[1]*v2[1] + v1[2]*v2[2];
-    return(result);
+
+    result =
+        v1[0] * v2[0]
+      + v1[1] * v2[1]
+      + v1[2] * v2[2]
+    ;
+    return (result);
 }
 
 void NormalizeVectorC(float *v)
 {
     register float len;
-    len = sqrtf(v[0]*v[0] + v[1]*v[1] + v[2]*v[2]);
-    if (len > 0.0f)
-    {
-        v[0] /= len;
-        v[1] /= len;
-        v[2] /= len;
-    }
+
+    len = v[0]*v[0] + v[1]*v[1] + v[2]*v[2];
+    if (len == 0.0f)
+        return;
+
+    len = sqrtf(len); /* len >= 0, because a*a + b*b + c*c is never negative. */
+    v[0] /= len;
+    v[1] /= len;
+    v[2] /= len;
 }
 
 void TransformVectorC(float *src, float *dst, float mat[4][4])
 {
-  dst[0] = mat[0][0]*src[0] + mat[1][0]*src[1] + mat[2][0]*src[2];
-  dst[1] = mat[0][1]*src[0] + mat[1][1]*src[1] + mat[2][1]*src[2];
-  dst[2] = mat[0][2]*src[0] + mat[1][2]*src[1] + mat[2][2]*src[2];
+    float products[4][4];
+    register int i, j;
+    const int limit = 4 - 1; /* either xmm[127..0] or just xmm[95..0] works */
+
+#if 0

```

Comment:
Perhaps, something like this?  https://github.com/cxd4/project64/commit/7d6ff5d29c2994a42d9782bf99d4dbc0648ba48e

The problem with SSE code is that even if it is faster (which it very well can be), it ironically can use up more space in the byte-code cache.  Even if this code is faster it can have a less direct effect outside of the functions--therefore in my opinion we should always default to optimizing algorithm for compact size and small binary builds, before testing with a profiler to see where the bottlenecks are and inverting the approach to maximum performance.  (Plus the #if 0 stuff helps people understand the loops...it's kind of easy to see something like (2x + 4x + 6x + 8x + ...) iterated within the C source, than only a loop on += 2*i*x or whatever.)